### ISSUE TITLE
[BUG] Gradient transform outputs are transposed compared to auto-diff framework

### ISSUE THREAD

**Issue by dime10** [PennyLaneAI/pennylane#2296]
### Expected behavior

When using a gradient transform directly to compute the Jacobian of a QNode, the output should be the same as that obtained via an auto-diff framework.

```py
dev = qml.device("default.qubit", wires=2)
x = np.array(0.543, requires_grad=True)

@qml.qnode(dev, diff_method="finite-diff")
def circuit(x):
    qml.RX(x, wires=[0])
    qml.RY(x, wires=[0])
    qml.CNOT(wires=[0, 1])
    return qml.probs(wires=0), qml.probs(wires=1)

transform_output = qml.gradients.finite_diff(circuit)(x)
framework_output = qml.jacobian(circuit)(x)

print(transform_output)
print(framework_output)
```

### Actual behavior

The output produced by the transform is the transpose of that produced by the framework (or vice-versa):

```
[[-0.44238498 -0.44238498]
 [ 0.44238498  0.44238498]]
[[-0.44238498  0.44238498]
 [-0.44238498  0.44238498]]
``` 

Besides `qml.gradients.finite_diff`, the same applies to other gradient transforms `qml.gradients.param_shift` and `qml.gradients.param_shift_cv` (harder to test as `qml.probs` is not supported). `qml.gradients.param_shift_hessian` does not appear to be affected, possibly due to the double application of the auto-differentiation.

### Additional information

This issue may only be noticeable when the QNode output is (at least) 2 dimensional.

For reference, the output is at least mostly consistent across `diff_method`s when and frameworks, with the exception of PyTorch which exhibits the transpose in `finite-diff` and `parameter-shift`, and produces a misshaped tensor for `backprop`:

```py
dev = qml.device("default.qubit", wires=2)
x = np.array(0.543, requires_grad=True)
x_torch = torch.tensor(x, dtype=torch.float64, requires_grad=True)
x_jax = jax.numpy.array(x)
x_tf = tf.Variable(x, dtype=tf.float64)
shapes, results = [], []

for m in ["backprop", "parameter-shift", "finite-diff"]:
    s, r = {}, {}

    @qml.qnode(dev, diff_method=m)
    def circuit(x):
        qml.RX(x, wires=[0])
        qml.RY(x, wires=[0])
        qml.CNOT(wires=[0, 1])
        return qml.probs(wires=0), qml.probs(wires=1)

    circuit.interface = "autograd"
    r["auto"] = qml.jacobian(circuit)(x)
    s["auto"] = qml.math.shape(r["auto"])

    circuit.interface = "jax"
    r["jax"] = jax.jacobian(circuit)(x_jax)
    s["jax"] = qml.math.shape(r["jax"])

    circuit.interface = "tf"
    with tf.GradientTape() as tape:
        y = circuit(x_tf)
    r["tf"] = tape.jacobian(y, x_tf).numpy()
    s["tf"] = qml.math.shape(r["tf"])

    circuit.interface = "torch"
    r["torch"] = torch.autograd.functional.jacobian(circuit, x_torch).numpy()
    s["torch"] = qml.math.shape(r["torch"])

    r['diff_method'] = m
    s['diff_method'] = m
    results.append(r)
    shapes.append(s)

header = ["diff_method", "auto", "jax", "tf", "torch"]
s_table = PrettyTable(field_names=header, align='l')
for s in shapes:
    s_table.add_row([s["diff_method"], s["auto"], s["jax"], s["tf"], s["torch"]])
r_table = PrettyTable(field_names=header, align='l')
for r in results:
    r_table.add_row([r["diff_method"], r["auto"], r["jax"], r["tf"], r["torch"]])
print(s_table)
print(r_table)
```

```
+-----------------+--------+--------+--------+--------+
| diff_method     | auto   | jax    | tf     | torch  |
+-----------------+--------+--------+--------+--------+
| backprop        | (2, 2) | (2, 2) | (2, 2) | (4,)   |
| parameter-shift | (2, 2) | (2, 2) | (2, 2) | (2, 2) |
| finite-diff     | (2, 2) | (2, 2) | (2, 2) | (2, 2) |
+-----------------+--------+--------+--------+--------+
+-----------------+-----------------------------+-----------------------------+-----------------------------+---------------------------------------------------+
| diff_method     | auto                        | jax                         | tf                          | torch                                             |
+-----------------+-----------------------------+-----------------------------+-----------------------------+---------------------------------------------------+
| backprop        | [[-0.44238494  0.44238494]  | [[-0.44238496  0.442385  ]  | [[-0.44238494  0.44238494]  | [-0.44238494  0.44238494 -0.44238494  0.44238494] |
|                 |  [-0.44238494  0.44238494]] |  [-0.44238496  0.442385  ]] |  [-0.44238494  0.44238494]] |                                                   |
| parameter-shift | [[-0.44238494  0.44238494]  | [[-0.44238493  0.44238493]  | [[-0.44238494  0.44238494]  | [[-0.44238494 -0.44238494]                        |
|                 |  [-0.44238494  0.44238494]] |  [-0.44238493  0.44238493]] |  [-0.44238494  0.44238494]] |  [ 0.44238494  0.44238494]]                       |
| finite-diff     | [[-0.44238498  0.44238498]  | [[-0.44238496  0.44238496]  | [[-0.44238498  0.44238498]  | [[-0.44238498 -0.44238498]                        |
|                 |  [-0.44238498  0.44238498]] |  [-0.44238496  0.44238496]] |  [-0.44238498  0.44238498]] |  [ 0.44238498  0.44238498]]                       |
+-----------------+-----------------------------+-----------------------------+-----------------------------+---------------------------------------------------+
```

### System information

```shell
-
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by trbromley** [comment]
@dime10 can we close this issue as resolved? Here is how it looks in the different interfaces:

## Autograd

It works, after adjustments for new return type style :heavy_check_mark: 

```python
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit", wires=2)
x = np.array(0.543, requires_grad=True)

@qml.qnode(dev, diff_method="finite-diff")
def circuit(x):
    qml.RX(x, wires=[0])
    qml.RY(x, wires=[0])
    qml.CNOT(wires=[0, 1])
    return qml.probs(wires=0), qml.probs(wires=1)

def cost(x):
    return np.stack(circuit(x))

transform_output = qml.gradients.finite_diff(circuit)(x)
framework_output = qml.jacobian(cost)(x)

print(transform_output)
print(framework_output)
```
```
(array([-0.44238498,  0.44238498]), array([-0.44238498,  0.44238498]))
[[-0.44238498  0.44238498]
 [-0.44238498  0.44238498]]
```

## Torch

It works, subject to a bug being fixed in #4882. :heavy_check_mark: 

```python
import pennylane as qml
import torch

dev = qml.device("default.qubit")
x = torch.tensor(0.543, requires_grad=True, dtype=torch.float64)

diff_method = "finite-diff"

@qml.qnode(dev, diff_method=diff_method)
def circuit(x):
    qml.RX(x, wires=[0])
    qml.RY(x, wires=[0])
    qml.CNOT(wires=[0, 1])
    return qml.probs(wires=0), qml.probs(wires=1)

transform_output = qml.gradients.finite_diff(circuit)(x)
framework_output = torch.autograd.functional.jacobian(circuit, x)

print(transform_output)
print(framework_output)
```
```
(tensor([-0.4424,  0.4424], dtype=torch.float64,
       grad_fn=<ReshapeAliasBackward0>), tensor([-0.4424,  0.4424], dtype=torch.float64,
       grad_fn=<ReshapeAliasBackward0>))
(tensor([-0.4424,  0.4424], dtype=torch.float64), tensor([-0.4424,  0.4424], dtype=torch.float64))
```

## JAX

It works :heavy_check_mark:. A complementary fix was done in #4879.

```python
import pennylane as qml
import jax

from jax import config
config.update("jax_enable_x64", True)

dev = qml.device("default.qubit")
x = jax.numpy.array(0.543)

@qml.qnode(dev, diff_method="finite-diff")
def circuit(x):
    qml.RX(x, wires=[0])
    qml.RY(x, wires=[0])
    qml.CNOT(wires=[0, 1])
    return qml.probs(wires=0), qml.probs(wires=1)

transform_output = qml.gradients.finite_diff(circuit)(x)
framework_output = jax.jacobian(circuit)(x)

jax.numpy.allclose(jax.numpy.stack(transform_output), jax.numpy.stack(framework_output))
```

## TensorFlow

It works :heavy_check_mark: 

```python
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit")

@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=[0])
    qml.RY(x, wires=[0])
    qml.CNOT(wires=[0, 1])
    return qml.probs(wires=0), qml.probs(wires=1)

def cost(x):
    return tf.stack(circuit(x))

x = tf.Variable(0.543)

with tf.GradientTape() as tape:
    transform_output = qml.gradients.param_shift(circuit)(x)
    res = cost(x)

framework_output = tape.jacobian(res, x)

print(transform_output)
print(framework_output)
```
```
(<tf.Tensor: shape=(2,), dtype=float64, numpy=array([-0.44238493,  0.44238493])>, <tf.Tensor: shape=(2,), dtype=float64, numpy=array([-0.44238493,  0.44238493])>)
tf.Tensor(
[[-0.44238496  0.44238496]
 [-0.44238496  0.44238496]], shape=(2, 2), dtype=float32)
```

Finally, here is the code you shared in the Additional Information section:

```python
from prettytable import PrettyTable
import pennylane as qml
from pennylane import numpy as np
import jax
import torch
import tensorflow as tf

dev = qml.device("default.qubit", wires=2)
x = np.array(0.543, requires_grad=True)
x_torch = torch.tensor(x, dtype=torch.float64, requires_grad=True)
x_jax = jax.numpy.array(x)
x_tf = tf.Variable(x, dtype=tf.float64)
shapes, results = [], []

for m in ["backprop", "parameter-shift", "finite-diff"]:
    s, r = {}, {}

    @qml.qnode(dev, diff_method=m)
    def circuit(x):
        qml.RX(x, wires=[0])
        qml.RY(x, wires=[0])
        qml.CNOT(wires=[0, 1])
        return qml.probs(wires=0), qml.probs(wires=1)

    def cost(x):
        return qml.math.stack(circuit(x))
    
    circuit.interface = "autograd"
    r["auto"] = qml.jacobian(cost)(x)
    s["auto"] = qml.math.shape(r["auto"])

    circuit.interface = "jax"
    r["jax"] = jax.numpy.stack(jax.jacobian(circuit)(x_jax))
    s["jax"] = qml.math.shape(r["jax"])

    circuit.interface = "tf"
    with tf.GradientTape() as tape:
        y = cost(x_tf)
    r["tf"] = tape.jacobian(y, x_tf).numpy()
    s["tf"] = qml.math.shape(r["tf"])

    circuit.interface = "torch"
    r["torch"] = torch.stack(torch.autograd.functional.jacobian(circuit, x_torch)).numpy()
    s["torch"] = qml.math.shape(r["torch"])

    r['diff_method'] = m
    s['diff_method'] = m
    results.append(r)
    shapes.append(s)

header = ["diff_method", "auto", "jax", "tf", "torch"]
s_table = PrettyTable(field_names=header, align='l')
for s in shapes:
    s_table.add_row([s["diff_method"], s["auto"], s["jax"], s["tf"], s["torch"]])
r_table = PrettyTable(field_names=header, align='l')
for r in results:
    r_table.add_row([r["diff_method"], r["auto"], r["jax"], r["tf"], r["torch"]])
print(s_table)
print(r_table)
```
```
+-----------------+--------+--------+--------+--------+
| diff_method     | auto   | jax    | tf     | torch  |
+-----------------+--------+--------+--------+--------+
| backprop        | (2, 2) | (2, 2) | (2, 2) | (2, 2) |
| parameter-shift | (2, 2) | (2, 2) | (2, 2) | (2, 2) |
| finite-diff     | (2, 2) | (2, 2) | (2, 2) | (2, 2) |
+-----------------+--------+--------+--------+--------+
+-----------------+-----------------------------+-----------------------------+-----------------------------+-----------------------------+
| diff_method     | auto                        | jax                         | tf                          | torch                       |
+-----------------+-----------------------------+-----------------------------+-----------------------------+-----------------------------+
| backprop        | [[-0.44238494  0.44238494]  | [[-0.44238496  0.442385  ]  | [[-0.44238494  0.44238494]  | [[-0.44238494  0.44238494]  |
|                 |  [-0.44238494  0.44238494]] |  [-0.44238496  0.442385  ]] |  [-0.44238494  0.44238494]] |  [-0.44238494  0.44238494]] |
| parameter-shift | [[-0.44238494  0.44238494]  | [[-0.4423849  0.4423849]    | [[-0.44238494  0.44238494]  | [[-0.44238494  0.44238494]  |
|                 |  [-0.44238494  0.44238494]] |  [-0.4423849  0.4423849]]   |  [-0.44238494  0.44238494]] |  [-0.44238494  0.44238494]] |
| finite-diff     | [[-0.44238498  0.44238498]  | [[-0.527364  0.527364]      | [[-0.44238498  0.44238498]  | [[-0.44238498  0.44238498]  |
|                 |  [-0.44238498  0.44238498]] |  [-0.527364  0.527364]]     |  [-0.44238498  0.44238498]] |  [-0.44238498  0.44238498]] |
+-----------------+-----------------------------+-----------------------------+-----------------------------+-----------------------------+
```

**Comment by dime10** [comment]
Hi @trbromley, great work! Looks good to me 👍
